### PR TITLE
Updated InternalLogger to Create Directories If Needed

### LINKED ISSUE #802: Internal Logger should create directories when needed
Hi,

I'm using Internal Logging with following NLog.config.

	<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" 
		xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	      internalLogFile="c:\Logs\InternalLog.log" 
	      internalLogLevel="Trace">
	...
	</nlog>

If the folder "c:\Logs" already exists, Internal Logging is appending.
But, if the folder doesn't exist, it not appending.

I hope insert this code before 
https://github.com/NLog/NLog/blob/4.0.1/src/NLog/Common/InternalLogger.cs#L323

`Directory.CreateDirectory(Path.GetDirectoryName(filename));`

The current NLog version: 4.0.1

Thanks,

### PR DESCRIPTION
**Pull request by kevindaub**
Updated the InternalLogger to create directories if they are needed.

Resolves #802

### CODE REVIEW

File: src/NLog/Common/InternalLogger.cs [hunk 1]
```diff
@@ -320,6 +320,7 @@ private static void Write(LogLevel level, string message, object[] args)
                 var logFile = LogFile;
                 if (!string.IsNullOrEmpty(logFile))
                 {
+                    CreateDirectoriesIfNeeded(logFile);
```
Comment (304NotModified) [inline src/NLog/Common/InternalLogger.cs]: Please move to constructor (performance)

File: tests/NLog.UnitTests/Common/InternalLoggerTests.cs [hunk 2]
```diff
@@ -336,6 +336,58 @@ public void TimestampTests()
                 Assert.Contains(expectedDateTime + ".", str);
             }
         }
+        
+        [Fact]
+        public void CreateDirectoriesIfNeeded()
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Common/InternalLoggerTests.cs]: Nice test :+1:

**Comment by 304NotModified** [comment]
Thanks! Please check the first note.

File: src/NLog/Common/InternalLogger.cs (hunk 1 above)
Comment (kevindaub) [inline src/NLog/Common/InternalLogger.cs]: Unfortunately, that will make my unit test moot, but agree with the performance.

**Comment by 304NotModified** [comment]
That's a lot of failed tests :/

**Comment by kevindaub** [comment]
That's a lot better.  I forgot to remove out the unit test.  Thoughts on moving all the static logic to an instance class that is initialized as a singleton.  I think that will give you the performance and the unit testibility.

**Comment by 304NotModified** [comment]
I really liked the unit test. I forgot that calling the method in a static ctor is difficult to mock. But we can fix this :)

New proposal:

- move the code to the setter (`InternalLogger.LogFile`)
- restore unit test

**Comment by 304NotModified** [comment]
I see the failing test. Dunno why they fail. Maybe you can change ` Assert.True` with `Assert.Equal` in those cases. (which is better)

**Comment by kevindaub** [comment]
The failed unit test was not my unit test.  However, I do notice that the tests in there seem to randomly fail.  It varies on my computer which one fails.  I take a closer to look to see if my change caused that.

**Comment by kevindaub** [comment]
My changes seems to have made the tests unstable.  It might be threading since it's a static class.  I'll dig in a bit more.

**Comment by 304NotModified** [comment]
yes, there are some tests which has race conditions. But that aren't those two failing tests. Do those tests successfully run on your local computer?

**Comment by kevindaub** [comment]
My test seems to make the race conditions worse. As long as my test doesn't
get executed, things are fine.  Perhaps making my unit test a skip?

On Monday, July 27, 2015, Julian Verdurmen <notifications@github.com> wrote:

> yes, there are some tests which has race conditions. But that aren't those
> two failing tests. Do those tests successfully run on your local computer?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NLog/NLog/pull/818#issuecomment-125111745>.
>


-- 
Sent from Gmail Mobile

**Comment by 304NotModified** [comment]
I think you should 'revert' the value of `InternalLogger.LogFile` at then end of your unit test.

**Comment by kevindaub** [comment]
Resetting the LogFile back to the previous value resolved the issue.  It worked repeatedly on my machine, which it was not doing before.  Thanks for the help.

**Comment by kevindaub** [comment]
Are we good with this pull request?

File: src/NLog/Common/InternalLogger.cs [hunk 3]
```diff
@@ -426,6 +443,21 @@ private static T GetSetting<T>(string configName, string envName, T defaultValue
...
+        
+        private static void CreateDirectoriesIfNeeded(string filename)
+        {
+            try
+            {
+                Directory.CreateDirectory(Path.GetDirectoryName(filename));
+            }
+            catch (Exception exception)
+            {
+                if (exception.MustBeRethrown())
```
Comment (304NotModified) [inline src/NLog/Common/InternalLogger.cs]: We need an log to the internallog here.

**Comment by 304NotModified** [comment]
Looks good! one small thing (because 'empty' catches are confusing)  - see the code note. Can you fix that?

File: src/NLog/Common/InternalLogger.cs (hunk 3 above)
Comment (kevindaub) [inline src/NLog/Common/InternalLogger.cs]: Do you mean something like this?

```csharp
            try
            {
                Directory.CreateDirectory(Path.GetDirectoryName(filename));
            }
            catch (Exception exception)
            {
                InternalLogger.Error("Cannot create needed directories {0}, Exception : {1}", Path.GetDirectoryName(filename), exception);
                if (exception.MustBeRethrown())
                {
                    throw;
                }
            }
```

**Comment by kevindaub** [comment]
Added a log if we are not rethrowing the exception.  But I think this actually might cause issues unless we log only to the console.  Whether we rethrow or not, the Write method uses the LogFile, which will probably not let that method finish.  Thoughts?

**Comment by 304NotModified** [comment]
You're worried about infinite loops? I don't think this is a problem, but maybe we should unit test it to be sure.

File: tests/NLog.UnitTests/Common/InternalLoggerTests.cs [hunk 4]
```diff
@@ -336,6 +336,67 @@ public void TimestampTests()
...
+                InternalLogger.Log(LogLevel.Info, "III");
+
+                AssertFileContents(tempFile, expected, Encoding.UTF8);
+                Assert.True(File.Exists(tempFile));
+            }
+            finally
+            {
+                // Reset LogFile to the previous value
+                InternalLogger.LogFile = previousLogFile;
+
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Common/InternalLoggerTests.cs]: please revert all the InternalLogger properties, to be sure. Maybe we should create a helper for that.

**Comment by kevindaub** [comment]
I did a small test myself and my fear was lacking any solid basis.  I did a small update.  Besides your last comment.  I think it's good.

**Comment by kevindaub** [comment]
Added an additional check for the failing test.  It looked like there was inner type that was auto generated by the compiler for the ReplaceLayoutRendererWrapper.  Running the test again with the additional check of t.IsPrivateNested showed that this was the only generated class.

**Comment by 304NotModified** [comment]
I will fix #839 soon and then we can use it in this PR.

**Comment by 304NotModified** [comment]
I recently enabled the Silverlight builds on appveyor.  Probably this will cause the build errors

**Comment by 304NotModified** [comment]
@kevindaub 

I created a separate PR for your fix. (https://github.com/NLog/NLog/pull/844). Thanks for the fix!

It should not lead to merge conflicts in this PR.

**Comment by kevindaub** [comment]
Great. Let me know if you need anything after the other pull request is
merged.

On Wednesday, August 12, 2015, Julian Verdurmen <notifications@github.com>
wrote:

> @kevindaub <https://github.com/kevindaub>
>
> I created a separate PR for your fix. (#844
> <https://github.com/NLog/NLog/pull/844>). Thanks for the fix!
>
> It should not lead to merge conflicts in this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NLog/NLog/pull/818#issuecomment-130457656>.
>


-- 
Sent from Gmail Mobile

**Comment by 304NotModified** [comment]
Yes I let you know. :)

it's almost fixed :)

**Comment by 304NotModified** [comment]
#844 isn't working as expected. So we don;t have to change something in this PR for now. .

**Pull request merged**